src/esp/physics/bullet/BulletArticulatedObject.cpp

```diff
@@ -714,5 +714,103 @@ void BulletArticulatedObject::updateKinematicState() {
   }
 }
 
+/**
+ * @brief Specific callback function for ArticulatedObject::contactTest to
+ * screen self-collisions.
+ */
+struct AOSimulationContactResultCallback
+    : public SimulationContactResultCallback {
+  btMultiBody* mb_ = nullptr;
+  btRigidBody* fixedBaseColObj_ = nullptr;
+
+  /**
+   * @brief Constructor taking the AO's btMultiBody as input to screen
+   * self-collisions.
+   */
+  AOSimulationContactResultCallback(btMultiBody* mb,
+                                    btRigidBody* fixedBaseColObj)
+      : mb_(mb), fixedBaseColObj_(fixedBaseColObj) {
+    bCollision = false;
+  }
+
+  /**
+   * @brief Called when a contact is detected.
+   *
+   * Sets a collision flag on every detected collision screening self-contacts
+   * if necessary.
+   * @param cp Contains detailed information about the contact point being
+   * added.
+   */
+  btScalar addSingleResult(
+      CORRADE_UNUSED btManifoldPoint& cp,
+      CORRADE_UNUSED const btCollisionObjectWrapper* colObj0Wrap,
+      CORRADE_UNUSED int partId0,
+      CORRADE_UNUSED int index0,
+      CORRADE_UNUSED const btCollisionObjectWrapper* colObj1Wrap,
+      CORRADE_UNUSED int partId1,
+      CORRADE_UNUSED int index1) override {
+    // screen self-collisions
+    if (!mb_->hasSelfCollision()) {
+      const btMultiBodyLinkCollider* mblc1 =
+          btMultiBodyLinkCollider::upcast(colObj0Wrap->getCollisionObject());
+      const btMultiBodyLinkCollider* mblc2 =
+          btMultiBodyLinkCollider::upcast(colObj1Wrap->getCollisionObject());
+      if (mblc1 && mblc2) {
+        if (mblc1->m_multiBody == mblc2->m_multiBody) {
+          // link self-collision
+          return 0;
+        }
+      } else if (colObj0Wrap->getCollisionObject() == fixedBaseColObj_) {
```

Comment:
Shouldn't "multibody link/base versus fixedBaseColObj" be already getting filtered out in the broadphase via use of CollisionGroups?

I'm not sure if getCollisionWorld()->contactTest uses the same CollisionGroup logic for collision filtering that stepSimulation uses. I thought it did.

Reply:
Good point, I dug into this a bit. We can configure the mask and group for the filter on a per-object basis. I added this change. However, we also need to separately handle self-collision between links. 

This also means that STATIC objects and fixed base will not report contact with other STATICs such as the stage by default. I added an option to override that behavior for common use cases. For example placing a fixed base ArticulatedObject in the stage via rejection sampling.